propolis/src/chardev/sock.rs

```diff
@@ -1,137 +1,117 @@
-use std::collections::VecDeque;
 use std::fs;
-use std::io::{ErrorKind, Read, Result, Write};
-use std::os::unix::io::AsRawFd;
-use std::os::unix::net::{UnixListener, UnixStream};
+use std::io::{ErrorKind, Result};
+use std::os::unix::net::UnixListener as StdUnixListener;
 use std::path::Path;
 use std::sync::{Arc, Condvar, Mutex, Weak};
 
-use crate::dispatch::events::{Event, EventTarget, FdEvents, Resource, Token};
-use crate::dispatch::DispCtx;
-use crate::util::self_arc::*;
+use crate::chardev::{Sink, Source};
+use crate::dispatch::Dispatcher;
 
-use super::{Sink, Source};
+use tokio::net::unix::SocketAddr;
+use tokio::net::{UnixListener, UnixStream};
+use tokio::sync::Notify;
 
-pub struct UDSock {
-    socks: Mutex<Socks>,
-    sink_driver: Mutex<SinkDriver>,
-    source_driver: Mutex<SourceDriver>,
-    cv: Condvar,
-    sa_cell: SelfArcCell<Self>,
-}
-struct Socks {
-    state: SockState,
-    server: UnixListener,
-    client: Option<UnixStream>,
-    client_token_fd: Option<Token>,
-}
+#[derive(Default)]
 struct SinkDriver {
-    sink: Option<Arc<dyn Sink<DispCtx>>>,
-    buf: VecDeque<u8>,
+    sink: Mutex<Option<Arc<dyn Sink>>>,
+    notify: Notify,
 }
-struct SourceDriver {
-    source: Option<Arc<dyn Source<DispCtx>>>,
-    buf: VecDeque<u8>,
-}
-
 impl SinkDriver {
-    fn new(bufsz: usize) -> Self {
-        assert!(bufsz > 0);
-        Self { sink: None, buf: VecDeque::with_capacity(bufsz) }
-    }
-}
-impl SourceDriver {
-    fn new(bufsz: usize) -> Self {
-        assert!(bufsz > 0);
-        Self { source: None, buf: VecDeque::with_capacity(bufsz) }
-    }
-}
-
-#[derive(Copy, Clone, PartialEq, Eq, Debug)]
-enum BufState {
-    Steady,
-    ProcessCapable,
-    ProcessRequired,
-}
-
-trait BufDriver {
-    fn drive(&mut self);
-    fn buffer_state(&self) -> BufState;
-
-    fn drive_transfer(&mut self) -> Option<BufState> {
-        let old = self.buffer_state();
-        self.drive();
-        let new = self.buffer_state();
-
-        if new != old {
-            Some(new)
-        } else {
-            None
+    async fn drive(&self, client: &UnixStream) -> Result<()> {
+        let sink = {
+            let guard = self.sink.lock().unwrap();
+            guard.as_ref().map(|s| Arc::clone(s))
+        };
+        if sink.is_none() {
+            return Ok(());
         }
-    }
-}
-
-impl BufDriver for SinkDriver {
-    fn drive(&mut self) {
-        if let Some(sink) = self.sink.as_ref() {
-            while let Some(b) = self.buf.pop_front() {
-                if !sink.sink_write(b) {
-                    self.buf.push_front(b);
-                    break;
+        let sink = sink.unwrap();
+        loop {
+            let mut buf = [0u8];
+            loop {
```

Comment:
Since we're already in an async context, can't we replace the nested loop with a `AsyncReadExt::read` call:

```rust
use tokio::io::AsyncReadExt;
...
if client.read(&mut buf).await? == 0 {
    return Ok(());
}
```

Reply:
Ah, that needs &mut to client, right.

Reply:
I'm iterating on changes to the chardev drivers to dogfood this async stuff.  Hopefully it'll address your thoughts there.